justniffer.8
.TH JUSTNIFFER 8 "March 22, 2025" "" ""
.SH NAME
\fBjustniffer \fP- tcp flow sniffer
\fB
.SH SYNOPSIS
.nf
.fam C
\fBjustniffer\fP  [ [\fB-i\fP \fBinterface\fP] or [\fB-f\fP <tcpdump file>] ] 
            [\fB-p\fP <packet filter>] 
            [\fB-m\fP]
            [\fB-u\fP or \fB-x\fP]
            [\fB-r\fP or \fB-l\fP <log format> or \fB-a\fP <log format>]
            [\fB-c\fP <config file>]
            [\fB-e\fP <external program>]
            [\fB-U\fP <user> ]
            [\fB-n\fP <not_found_string> ]
            [\fB-d\fP <max concurrent hosts for fragmented ip> ]
            [\fB-s\fP <max concurrent tcp streams> ]
            [\fB-F\fP]  
.fam T

\fBExamples:\fP 
  Logging network traffic in Apache like format:
      \fBjustniffer -i eth0\fP 

  Tracing tcpdump dump in Apache like format:
      \fBjustniffer -f file.cap\fP 
  
  Tracing tcp streams from the network:
      \fBjustniffer -i eth0 -r\fP 

  Logging network traffic in Apache like format appending response times:
      \fBjustniffer -i eth0 -a " %response.time"\fP 

  Logging network traffic in a customized format:
      \fBjustniffer -i eth0 -l "%source.ip%tab%dest.ip%tab%request.header.host%tab%response.time"\fP 


.fi
.SH DESCRIPTION
\fBjustniffer\fP captures reassembles and reorders TCP packets, performs IP packet defragmentation and displays the tcp
.br
flow or part of it. It is useful for logging network traffic in a 'standard' (web server like) or in a customized way.
.br
It can log network services performances (e.g. web server response time, application server keep alive behaviour, etc.).
.br
Output format can be customized using the log format option \fB-l\fP (see \fBFORMAT KEYWORDS\fP). Most of them can be
.br
used for retrieving HTTP protocol informations. See \fBEXAMPLES\fP.


.SH TRACKING PERFORMANCES

 The main feature of \fBjustniffer\fP is to analize network traffic to monitor performances. 
.br
 The perfomance related keywords are:
  %session.time
    %connection.time
    %idle.time.0
    %request.time
    %response.time
        %response.time.begin
        %response.time.end
    %idle.time.1

 +---------+                           +---------+
 |         |                           |         |
 |  Client |                           | Server  |
 |         |                           |         |
 +---------+                           +---------+
      |                                     |
      |  -----   connect syn   -------->    |----+--------------------------------------+
      |                                     |    |                                      |
      |  <------   syn/ack    --------->    |    | %connection.time                     |
      |                                     |    |                                      |
      |  -------     ack     ---------->    |    |                                      |
      |           ESTABLISHED               |----+                                      |
      |                                     |    | %idle.time.0                         |
      |                                     |    |(after connection, before             |
      |                                     |    | request)                             |
      |                                     |    |                                      |
      |  ---  request/first packet  --->    |----+--------------------------------------+
      |                                     |    |                                      |
      |                                     |    |                                      |
      |  ---  request/....          --->    |    | %request.time                        |
      |  <------     ack     -----------    |    |                                      |
      |                                     |    |                                      |
      |  ---   request/last packet  --->    |    |                                      | 
      |  <------     ack     -----------    |----+--------------------+-----------------+ %session.time
      |                                     |    |                    |                 |
      |                                     |    |                    |                 |
      |                                     |    |%reponse.time.begin |                 |
      |                                     |    |                    |                 |
      |  <--  response/first packet ----    |----+                    | reponse         |
      |  -------     ack     ---------->    |    |                    | time            |
      |                                     |    |                    |                 |
      |  <--  response/....         ----    |    |%reponse.time.end   |                 |
      |  -------     ack     ---------->    |    |                    |                 |
      |                                     |    |                    |                 |
      |  <--  response/last packet  ----    |    |                    |                 |
      |  -------     ack     ---------->    |----+--------------------+-----------------+
      |                                     |    |                                      |
      |                                     |    |                                      |
      |                                     |    | %idle.time.1 (after response,        |
      |                                     |    | before new request or close)         |
      |                                     |    |                                      |
      |  <------   close      --------->    |----+--------------------+-----------------+
      |                                     |    |
      |                                     |    |


.SH OPTIONS
\fB-i\fP or \fB--interface\fP=<\fBinterface\fP>\fBinterface\fP to listen on (e.g. eth0, en1, etc.)
.TP
Example: justniffer -i eth0
.TP
.B
\fB-f\fP or \fB--filecap\fP=<file>
tcpdump file to read from (for offline network traffic processing).
.br
It must be a pcap file (produced by network capture programs such as tcpdump or wireshark)
.br
\fBWARNING\fP: justniffer needs a complete dump, usually, sniffers collect just the few first (96) bytes per packet.
.br
(when using tcpdump you must specify "-s 0" option. Example: tcpdump -i eth0 -s 0 -w /tmp/file.cap)
.TP
Example: justniffer -f /tmp/file.cap
.TP
.B
\fB-F\fP or \fB--force-read-pcap\fP
Force the reading of a pcap file also if captured with truncated packets 
.br
(for example when using tcpdump without the "-s 0" option)
.TP
.B
\fB-m\fP or \fB--capture-in-the-middle\fP
Captures and reconstructs TCP streams in the middle (even without the initial connection).
.br
\fBWARNING\fP: it may yield unexpected results.
.TP
.B

\fB-p\fP or \fB--packet-filter\fP=<filter>
packet filter (pcap filter syntax) (see \fBpcap-filter\fP(7))
.TP
Examples: 

  sniffing proxy traffic
    \fBjustniffer -i eth0 -p "tcp port 8080"\fP

  sniffing row telnet traffic
    \fBjustniffer -i eth0 -r -p "tcp port 23"\fP

  sniffing row pop3 traffic
    \fBjustniffer -i eth0 -r -p "tcp port 110"\fP
    
  sniffing  traffic from/to a specific host 
    \fBjustniffer -i eth0 -r -p "host 10.10.10.2"\fP

  sniffing  HTTP traffic from/to a specific host and port 
    \fBjustniffer -i eth0 -r -p "host 10.10.10.2 and tcp port 80"\fP
.TP
.B
\fB-l\fP or \fB--log-format\fP=<format>
log format. You can specify the output string format containing reserved keyword that will be replaced with the proper value 
.br
(see \fBFORMAT KEYWORDS\fP). 
.br
If missing, the CLF (Common Log Format, used by  almost all web servers) is used as default.
.br
"%source.ip - - [%request.timestamp(%d/%b/%Y:%T %z)] \"%request.line\" %response.code %response.header.content-length \"%request.header.referer\" \"%request.header.user-agent)\""
.TP
Example: justniffer -i eth0 -l "\\"%request.line\\"%tab%response.time"

    "POST /v2/rss/alerts?src=ffbmext2.1.034 HTTP/1.1" 0.139011
    "POST /v2/rss/network/oreste.notelli?src=ffbmext2.1.034 HTTP/1.1" 0.623382
    "GET /man_page_howto.html HTTP/1.1"       0.024437
    "GET /ig?hl=en HTTP/1.1"  0.764945
    "GET /?hl=en&tab=wv HTTP/1.1"   0.242342
    "GET /s/-yCdCsgUnsI/css/homepage_c.css HTTP/1.1"        0.071942
    "GET /vi/YUvWcegtqik/default.jpg HTTP/1.1"      0.821472

.TP
.B
\fB-a\fP or \fB--append-log-format\fP=<format>
append log format. append the specified format to the default common log format (apache access_log)
.TP
Example: justniffer -i eth0 -a " %response.time"

192.168.10.102 - - [25/Jun/2009:12:02:18 +0200] "GET /nwshp?hl=en&client=firefox-a&rls=com.ubuntu:en-US:unofficial&hs=Ap6&tab=wn HTTP/1.1" 200 94077 "http://www.google.it/webhp?hl=en&safe=off&client=firefox-a&rls=com.ubuntu:en-US:unofficial&hs=Ap6&btnG=Search" "Mozilla/5.0 (X11; U; Linux i686; en-US;rv:1.9.0.11) Gecko/2009060308 Ubuntu/9.04 (jaunty) Firefox/3.0.10)" 0.712165

.TP
.B
\fB-r\fP or \fB--raw\fP
show raw stream. it is a shortcut for  -l %request%response
.TP
Example: justniffer -i eth0 -r
.TP
.B
\fB-s\fP or \fB--max-tcp-streams\fP
max concurrent tcp stream. (default= 65536) excess will be discarded
.TP
.B
\fB-d\fP or \fB--max-fragmented-ip\fP
max concurrent fragmented ip. (default= 65536) excess will be discarded
.TP
.B
\fB-x\fP or \fB--hex-encode\fP
encode unprintable characters in [<char hexcode>] format
.TP
Example: 
  justniffer -i eth0 -rx
.TP
.B
\fB-u\fP or \fB--unprintable\fP
encode as dots (.) unprintable characters ( for more control on character you should use pipilining to cat -v. 
.TP
Example: 
  justniffer -i eth0 -l "%request%newline%response"| cat -v
  justniffer -i eth0 -ru
.TP
.B
\fB-e\fP or \fB--execute\fP=<external program>
call the specified external program/shell script pipelining the standard output for each request/response phase
You can write shell script for handling, for example, HTTP traffic
.TP
Example: 
  justniffer -i eth0 -l "%request%newline%response" -e "tail -2 "

.TP
.B
\fB-n\fP or \fB--not-found\fP=<not found string>
Not found string. It is used to replace a specified keyword when it cannot be valued because it is not found.
All request.header.* and response.header.* keywords can override the "not found string" passing it as parameter.
For example: 
 %request.header.host() will be replaced by the Host header value, or an empty string if Host header not found
 %request.header.host(UNKNOWN) will be replaced by the Host header value, or the UNKNOWN string if Host header not found
 %request.header.host will be replaced by the Host header value, or , if Host header not found, with the string specified by the \fB-n\fP option 
.TP
Examples: 
  justniffer -i eth0 -l "%request.header.connection" -n N/A  

  will produce such logs:
    N/A
    N/A
    keep-alive
    close

Note: each keyword can override the "not found string" value:

  justniffer -i eth0 -l "%request.header.connection(None) %request.header.connection(-) %request.header.connection() %request.header.connection" -n N/A  

  will produce such logs:
    None -  N/A
    None -  N/A
    keep-alive keep-alive keep-alive keep-alive
    close close close close
.TP
.B
\fB-U\fP or \fB--user\fP=<user>
User to imperfonificate when executing the program specified with the \fB-e\fP option, used to avoid to security exploits when running justniffer with root privileges
.TP
Example: 
  justniffer -i eth0 -l "%request%newline%response" -e " grep password >> /tmp/passwords.txt"  -U guest
.TP
.B
\fB-c\fP or \fB--config\fP=<config file>
configuration file. You can specify options in a configuration file (command line options override file configuration options) using the following format specifications:
.PP
.nf
.fam C
       <option> = <value>

       \fBconfiguration file example\fP:
.fam T
.fi
.RS
log-format = "%request.url %request.header.host %response.code %response.time"
.PP
packet-filter = "tcp port 80 or tcp port 8080 or tcp port 3526"


.SH FORMAT KEYWORDS
List of all recognized keywords:
.RS
.TP
.B
%close.originator([not applicable string])
who closed the connection: client, server, timedout, n/a
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%close.time
elapsed time from last response and when the connection is closed
.TP
.B
%close.timestamp([format])
is replaced by the close timestamp. You can use optional format specification (see \fBTIMESTAMP FORMAT\fP) 
.TP
.B
%connection.timestamp2([not applicable string])
is replaced by the close timestamp using format "seconds.microseconds". 
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%connection
 connection persistence indicator:
    unique: the request/response is the unique in the tcp connection
    start: the request/response is the first in the tcp connection
    last : the request/response is the last in the tcp connection
    continue : the request/response is the middle in the tcp connection
.TP
.B
%connection.time([not applicable string])
elapsed time for the estabilishind the tcp connection
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%connection.timestamp([format])
is replaced by the connection timestamp. You can use optional format specification (see \fBTIMESTAMP FORMAT\fP) 
.TP
.B
%connection.timestamp2([not applicable string])
is replaced by the connection timestamp using format "seconds.microseconds"
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%idle.time.0([not applicable string])
elapsed time form when the connection is established and the request is started
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%idle.time.1([not applicable string])
elapsed time form when the last response and the next request (or the connection closing)
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%dest.ip
is replaced by the destination ip address
.TP
.B
%dest.port
is replaced by the destination tcp port
.TP
.B
%source.ip
is replaced by the source ip address
.TP
.B
%source.port
is replaced by the source tcp port
.TP
.B
%request
is replaced by the  the whole request ( (it is multiline and may contain unprintable characters)
.TP
.B
%request.timestamp([format])
is replaced by the request timestamp. You can use optional format specification (see \fBTIMESTAMP FORMAT\fP)
.TP
.B
%request.timestamp2([not applicable string])
is replaced by the request timestamp using format "seconds.microseconds"
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.size
is replaced by the request size (including request header size)
.TP
.B
%request.line
is replaced by the request line (e.g. "GET /index.html HTTP/1.1")
.TP
.B
%request.method
is replaced by the request method (e.g. GET, POST, HEAD)
.TP
.B
%request.url
is replaced by the url
.TP
.B
%request.protocol
is replaced by the protocolo (e.g. HTTP/1.0, HTTP/1.1) 
.TP
.B
%request.grep(<regular-expression>)
is replaced by the result of the specified regular expression applied on the whole request [Perl regular expression syntax, see \fBperlre\fP(1) or \fBperl\fP(1)]. The most nested subgroup is returned 
.TP
.B
%request.body
is replaced by the request body
.TP
.B
%request.header
is replaced by the request header (it is multiline)
.TP
.B
%request.header.authorization([not found string])
is replaced by the request Authorization header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.host([not found string])
is replaced by the request Host header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.user-agent([not found string])
is replaced by the request User-Agent header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.accept([not found string])
is replaced by the request Accept header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.accept-language([not found string])
is replaced by the request Accept-Language header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.accept-charset([not found string])
is replaced by the request Accept-Charset header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.accept-encoding([not found string])
is replaced by the request Accept-Encoding header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.connection([not found string])
is replaced by the request Connection header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.content-encoding([not found string])
is replaced by the request Content-Encoding header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.content-language([not found string])
is replaced by the request Content-Language header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.content-length([not found string])
is replaced by the request Content-Length header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.content-md5([not found string])
is replaced by the request Content-MD5 header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.cookie([not found string])
is replaced by the request Cookie header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.range([not found string])
is replaced by the request Range header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.referer([not found string])
is replaced by the request Referer header  value
The optional "" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.keep-alive([not found string])
is replaced by the request Keep-Alive header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.via([not found string])
is replaced by the request Via header  value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%request.header.value(<header-name>)
is replaced by the request header value (e.g. "%request.header.value(Cookie)")
.TP
.B
%request.header.grep(<regular-expression>)
is replaced by the result of the specified regular expression applied on the request header [Perl regular expression syntax, see \fBperlre\fP(1) or \fBperl\fP(1)]. The most nested subgroup is returned (e.g. to obtain the request URL:  "%request.header.grep(^[^\\s]*\\s*([^\\s]*))"
.TP
.B
%response
is replaced by the while response (it is multiline and may contain unprintable characters)
.TP
.B
%response.size
is replaced by the response size (including response the header size)
.TP
.B
%response.timestamp([format])
is replaced by the response timestamp. You can use optional format specification (see \fBTIMESTAMP FORMAT\fP)
.TP
.B
%response.timestamp2([not applicable string])
is replaced by the response timestamp using format "seconds.microseconds"
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.time([not applicable string])
is replaced by the response time (difference from the request time and the time the response finish)
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.time.begin([not applicable string])
elapsed time form when the request and the response start
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.time.end([not applicable string])
elapsed time form the response start and the response end
the  "not applicable" string is replaced in case the keyword value cannot be applicable. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.line
is replaced by the reponse line
.TP
.B
%response.protocol
is replaced by the reponse protocol
.TP
.B
%response.code
is replaced by the response code (e.g. 200, 404, 500, etc.)
.TP
.B
%response.body
is replaced by the response body
.TP
.B
%response.message
is replaced by response message (e.g. OK, Not Found, Internal Server Error, etc.)
.TP
.B
%response.grep(<regular-expression>)
is replaced by the result of the specified regular expression applied on the whole response [Perl regular expression syntax, see \fBperlre\fP(1) or \fBperl\fP(1)]. The most nested subgroup is returned 
.TP
.B
%response.header
is replaced by the response header (it is multiline)
.TP
.B
%response.header.accept-ranges([not found string])
is replaced by the response Accept-Ranges header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.allow([not found string])
is replaced by the response Allow header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.server([not found string])
is replaced by the response Server header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.date([not found string])
is replaced by the response Date header value 
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.content-language([not found string])
is replaced by the respone Content-Language header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.content-length([not found string])
is replaced by the respone Content-Lenght header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.content-md5([not found string])
is replaced by the respone Content-MD5 header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.content-type([not found string])
is replaced by the response Content-Type header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.content-encoding([not found string])
is replaced by the response Content-Encoding header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.content-language([not found string])
is replaced by the response Content-Language header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.transfer-encoding([not found string])
is replaced by the response Transfer-Encoding header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B

%response.header.expires([not found string])
is replaced by the response Expires header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.etag([not found string])
is replaced by the response ETag header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.cache-control([not found string])
is replaced by the response Cache-Control header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.last-modified([not found string])
is replaced by the response Last-Modified header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.pragma([not found string])
is replaced by the response Pragma header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.age([not found string])
is replaced by the response Age header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.connection([not found string])
is replaced by the response Connection header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.keep-alive([not found string])
is replaced by the response Keep-Alive header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.via([not found string])
is replaced by the response Via header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.vary([not found string])
is replaced by the response Vary header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.www-authenticate([not found string])
is replaced by the response WWW-Authenticate header value
The optional "not found" string is replaced in case the keyword value was not found. 
if not provided the -n value or the default value "-" is used 
.TP
.B
%response.header.set-cookie([not found string])

ChangeLog
2010-03-27  Oreste Notelli
      * fixed libnids for gcc4.4

2009-07-03  Oreste Notelli
      * forced flush on standard output whenever an new line is written 
      * close_originator handler fixed
      * lintian errors fixed

2009-06-29  Oreste Notelli
      * improved performances keywords adding a "not applicable string" settable by optional parameter
      * documentation updated for enhancement provided from version (0.5.3)

2009-06-26  Oreste Notelli
      * fixed request.header.* and response.header.* keywords bug
      * fixed help options
      * added connection.timestamp2 request.timestamp2 response.timestamp2 keywords

2009-06-25  Oreste Notelli
      * fixed %source.ip , %dest.ip , %source.port, %dest.port
      * fixed %idle.time.0 formatting
      * fixed %close.time formatting
      * added -n option
      * added -x encoding for unprintable chars
      * added "not found string" for request.header.* and response.header.* keywords

2009-06-01  Oreste Notelli
      * fixed default log format

2009-06-01  Oreste Notelli

      * fixed compilation issues with gcc 3.3.3
      * added connection.timestamp
      * fixed idle.time.0 formatting
      * fixed documentation
      * more adherence to RFC 2616: looking for headers in case insesitive mode
      * new HTTP headers added

2009-05-27  Oreste Notelli

      * added new extension scripts for saving http captured contents

2009-05-23  Oreste Notelli 

      * debianization

2009-04-21  Oreste Notelli 

      * script extensions option
      
      * grabbing http example


2007-09-18  Oreste Notelli  <oreste.notelli@plecno.com>

	* configure.in: version 1.2

	* src/formatter.cpp:
	don't print log for connection open and closed without data transfer



src/justniffer.py
import sys
sys.argv=["justiffer"]

import _justniffer
from collections import OrderedDict

class dict_object(dict):
    def __getattr__(self, name):
        return self[name]
    
    def __dir__(self):
        return self.keys()

BaseHandler = _justniffer.BaseHandler
class BaseImpl(BaseHandler):

    def __calc_time(self, time1, time0):
            values = self.values
            if values.has_key(time1) and values.has_key(time0):
                return values[time1]-values[time0]
    
    def _guess_protocol(self, request, response):
        try:
            part = response[:4]
            part.decode("ascii")
            prot =  part.strip().upper()
        except:
            prot = "".join(["\\"+hex(ord(p)) for p in part])
        return prot
    
    def __update_string_value(self, name, value):
        values = self.req_resp
        
        if not values.has_key(name):
            values[name]=value
        else:
            values[name]+=value
        
    def __init__(self):
        BaseHandler.__init__(self)
        self.values = dict_object()
        self.req_resp = dict_object()
        self.__onRequest_start_time=None
        self.__onRespone_start_time=None

    def onClose(self, stream, time):
        self.values["close_end_time"] = time
        
    def onOpening(self, stream, time):
        values = self.values
        values["source_ip"] = stream.src_ip
        values["source_port"] = stream.src_port
        values["dest_ip"] = stream.dst_ip
        values["dest_port"] = stream.dst_port
        self.values["opening_end_time"] = time
        
    def onOpen(self, stream, time):
        values = self.values
        values["source_ip"] = stream.src_ip
        values["source_port"] = stream.src_port
        values["dest_ip"] = stream.dst_ip
        values["dest_port"] = stream.dst_port
        self.values["open_end_time"] = time
        
    def onExit(self, stream):
        pass
        
    def onRequest(self, stream, time):
        values = self.values
        values["source_ip"] = stream.src_ip
        values["source_port"] = stream.src_port
        values["dest_ip"] = stream.dst_ip
        values["dest_port"] = stream.dst_port
        if (not values.has_key("request_start_time")):
            values["request_start_time"] = time
        values["request_end_time"] = time
        self.__update_string_value("request", stream.client_data)
        
    def onResponse(self, stream, time):
        values = self.values
        if (not values.has_key("response_start_time")):
            values["response_start_time"] = time
        values["response_end_time"] = time
        self.__update_string_value("response", stream.server_data)

    def __get_val(self, name, time):        
        if self.values.has_key(name):
            val = self.values[name]
        else:
            val = None
        return val
        
    def __get_reqrep_size(name):        
        def __call__(self, name_, time):
            if self.req_resp.has_key(name):
                val = len(self.req_resp[name])
            else:
                val = 0
            return val
        return __call__
        
    def __get_protocol(self, name, time):
        values = self.req_resp
        if values.has_key("response") and values.has_key("request"):
            protocol = self._guess_protocol(values.request,values.response)
        else:
            protocol=None
        return protocol
    
    def __ct(time1, time0):
        def __call__(self, name, time):
            return self.__calc_time(time1, time0)
        return __call__
    
    def __calc_idle_time_1(self, name, time):
        values = self.values
        idle_time_1=None
        if values.has_key("response_end_time"):
            if values.has_key("close_end_time"):
                idle_time_1 = self.__calc_time ("close_end_time" , "response_end_time")
            elif values.has_key("response_end_time"):
                idle_time_1 =  time - values.response_end_time
        return idle_time_1
    
    def __type_in_stream(self, name, time):
        values = self.values
        connection_time = values.connection_time
        close_time = values.close_time
        first = True if connection_time else False 
        last = True if close_time else False 
        continue_  = not (first and last)
        unique  = first and last
        if unique:
            type_="unique"
            
        elif continue_:
            type_="continue"
        elif first:
            type_="first"
        elif last:
            type_="last"
        else:
            type_="unknown"
        return type_

    @classmethod
    def on_exit(cls):
        pass
        
    od = OrderedDict
    attributes = od((
        ("response_size",__get_reqrep_size("response")),
        ("request_size",__get_reqrep_size("request")),
        ("protocol",__get_protocol),
        ("response_time",__ct("response_end_time" ,"request_end_time")),
        ("connection_time",__ct ("open_end_time" , "opening_end_time")),
        ("close_time",__ct ("close_end_time" , "response_end_time")),
        ("req_resp_time",__ct ("response_end_time" , "request_start_time")),
        ("idle_time_0",__ct ("response_start_time" , "request_end_time")),
        ("idle_time_1",__calc_idle_time_1),
        ("type_in_stream",__type_in_stream))
    )

    def append(self, outstream, time):
        values = self.values
        for attribute_name, func in self.attributes.items():
            attribute = func(self, attribute_name, time)
            values[attribute_name]= attribute
        self._on_complete(values, time)

    def _on_complete(self, values, time):
        pass

class Test(BaseImpl):
    def _on_complete(self, values, time):
        print values
    
    
def on_exit():
    print "on_exit"
